peas/sensors.py

```diff
@@ -125,45 +90,103 @@ def capture(self, use_mongo=True, send_message=True):
-            except TypeError:
-                self.logger.warning('Bad JSON: {0}'.format(sensor_value))
             except Exception as e:
-                self.logger.warning('Bad JSON: {0}'.format(sensor_value))
+                self.logger.warning('Exception while reading from sensor {}: {}', sensor_name, e)
 
-            if use_mongo and len(sensor_data) > 0:
-                if self.db is None:
-                    self.db = PanMongo()
-                    self.logger.info('Connected to PanMongo')
-                self.db.insert_current('environment', sensor_data)
-        else:
-            self.logger.debug('No sensor data received')
+        if use_mongo and len(sensor_data) > 0:
+            if self.db is None:
+                self.db = PanMongo()
+                self.logger.info('Connected to PanMongo')
+            self.db.insert_current('environment', sensor_data)
 
         return sensor_data
+
+
+def auto_detect_arduino_devices(comports=None, logger=None):
+    if comports is None:
+        comports = find_arduino_devices()
+    if not logger:
+        logger = get_root_logger()
+    result = []
+    for port in comports:
+        v = auto_detect_port(port, logger)
+        if v:
+            result.append(v)
+    return result
+
+
+def find_arduino_devices():
+    """Find devices (paths or URLs) that appear to be Arduinos.
+
+    Returns:
+        a list of strings; device paths (e.g. /dev/ttyACM1) or URLs (e.g. rfc2217://host:port
+        arduino_simulator://?board=camera).
+    """
+    comports = list_comports()
+    return [p.device for p in comports if 'Arduino' in p.description]
+
+
+def auto_detect_port(port, logger):
+    """Open a port and determine which type of board its producing output.
+
+    Returns: (name, serial_reader) if a recognizable device is connected, else None.
+    """
+    logger.debug('Attempting to connect to serial port: {}'.format(port))
+    try:
+        serial_reader = SerialData(port=port, baudrate=9600)
+        serial_reader.connect()
+        logger.debug('Connected to {}', port)
+    except Exception as e:
+        logger.warning('Could not connect to port: {}'.format(port))
+        return None
+    try:
+        reading = serial_reader.get_and_parse_reading()
+        if not reading:
+            return None
+        (ts, data) = reading
+        if isinstance(data, dict) and 'name' in data and isinstance(data['name'], str):
+            result = (data['name'], serial_reader)
+            serial_reader = None
+            return result
+        logger.warning('Unable to find board name in reading: {!r}', reading)
+        return None
+    except Exception as e:
+        logger.error('Exception while auto-detecting port {!r}: {!r}'.format(port, e))
+    finally:
+        if serial_reader:
+            serial_reader.disconnect()
+
+
+# Support testing by just listing the available devices.
+if __name__ == '__main__':
```

Comment:
FWIW, and regarding comment above about separate utils for discovering devices, I have mostly been making the module callable via `if __main__` _unless_ they are in the utils folder. However, that would make development a little annoying as you would have to run a util file to test the sensors file.  Still, could end up looking something like:
```
python pocs/utils/devices.py --list_arduinos
python pocs/utils/devices.py --list_cameras
python pocs/utils/devices.py --list_attached_hardware
```
etc.